File: app/src/main/java/com/openclassrooms/hexagonal/games/data/repository/PostRepository.java
package com.openclassrooms.hexagonal.games.data.repository;

import java.util.List;
import javax.inject.Inject;
import javax.inject.Singleton;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.openclassrooms.hexagonal.games.data.service.PostApi;
import com.openclassrooms.hexagonal.games.domain.model.Post;

/**
 * This class provides a repository for accessing and managing Post data.
 * It utilizes dependency injection to retrieve a PostApi instance for interacting
 * with the data source. The class is marked as a Singleton using @Singleton annotation,
 * ensuring there's only one instance throughout the application.
 */
@Singleton
public final class PostRepository
{

  private final PostApi postApi;

  /**
   * Constructor for PostRepository. Injects the PostApi dependency.
   *
   * @param postApi An instance of the PostApi to be used.
   */
  @Inject
  public PostRepository(PostApi postApi)
  {
    this.postApi = postApi;
  }

  /**
   * Retrieves a LiveData object containing a list of Posts ordered by creation date
   * in descending order.
   *
   * @return LiveData containing a list of Posts.
   */
  public LiveData<List<Post>> getPosts()
  {
    return new MutableLiveData<>(postApi.getPostsOrderByCreationDateDesc());
  }

  /**
   * Adds a new Post to the data source using the injected PostApi.
   *
   * @param post The Post object to be added.
   */
  public void addPost(Post post)
  {
    postApi.addPost(post);
  }

}

File: app/src/main/java/com/openclassrooms/hexagonal/games/data/repository/PostRepository.kt
package com.openclassrooms.hexagonal.games.data.repository

import com.openclassrooms.hexagonal.games.data.service.PostApi
import com.openclassrooms.hexagonal.games.domain.model.Post
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository class responsible for handling post-related operations.
 * This class abstracts the data source (Firebase) and provides methods to interact with posts.
 *
 * @property postApi The API service used to fetch and store post data.
 */
@Singleton
class PostRepository @Inject constructor(
    private val postApi: PostApi
) {

    /**
     * Retrieves a list of posts ordered by their creation date.
     *
     * @return A Flow emitting the result of the fetch operation.
     */
    fun getPosts(): Flow<PostResult> = postApi.getPostsOrderByCreationDate()

    /**
     * Adds a new post to the data source.
     *
     * @param post The post object containing details such as title, description, and timestamp.
     * @return A Flow emitting the result of the add operation.
     */
    fun addPost(post: Post): Flow<PostResult> = postApi.addPost(post)

    /**
     * Uploads a photo associated with a post to cloud storage.
     *
     * @param post The post containing the photo URL.
     * @return A Flow emitting the result of the upload operation.
     */
    fun addPhoto(post: Post): Flow<PostResult> = postApi.addPhoto(post)

    /**
     * Retrieves a specific post by its unique identifier.
     *
     * @param postId The unique identifier of the post to retrieve.
     * @return A Flow emitting the result of the fetch operation.
     */
    fun getPost(postId: String) : Flow<PostResult> = postApi.getPost(postId)

}